File: geohash.R
library(dplyr)
library(readr)
library(geohashTools)
library(leaflet)

# Read in example data and province shape files
df <- read_csv("ACLED.csv") %>% select(event_date, latitude, longitude)

# Geohash the data
df$hash <- gh_encode(df$latitude, df$longitude, precision = 4)
density <- df %>% group_by(hash) %>% count()
density[,c("hash_lat","hash_lon","lat_error","lon_error")] <- geohashTools::gh_decode(density$hash, include_delta = TRUE)
density <- density %>% mutate(
  lon_low = hash_lon - lon_error, lon_high = hash_lon + lon_error,
  lat_low = hash_lat - lat_error, lat_high = hash_lat + lat_error)

# Create a leaflet map
map_center <- tibble(lat = as.numeric(names(sort(table(round(density$hash_lat,1)),
                                                 decreasing = TRUE))[1]),
                     lon = as.numeric(names(sort(table(round(density$hash_lon,1)),
                                                 decreasing = TRUE))[1]))
log_n <- log(density$n)
pal <- colorNumeric(palette = "Reds", domain = range(log_n))
map_colors <- pal(log_n)
map <- leaflet() %>%
  setView(lng = map_center$lon-0.2, lat = map_center$lat+0.05, zoom = 6) %>%
  addProviderTiles(providers$Stamen.Toner) %>%
  addRectangles(
    lng1 = density$lon_low, lat1 = density$lat_low,
    lng2 = density$lon_high, lat2 = density$lat_high,
    fillColor = map_colors, fillOpacity = 0.7, stroke = FALSE,
    popup = as.character(paste("Geohash:", density$hash, "<br>",
                               "Total Hits:", density$n)))
map

File: geohash.py
import pandas as pd
import geopandas as gpd
import pygeohash as gh
import folium
from shapely import geometry
import math

# Read in example data and province shape files
df = pd.read_csv("ACLED.csv")
df = df[['event_date', 'latitude', 'longitude']]

# Geohash the data
df['hash'] = df.apply(lambda x: gh.encode(x.latitude, x.longitude, precision = 4), axis = 1)
density = df['hash'].value_counts().reset_index()
density.columns = ['hash', 'count']

def geohash_to_polygon(geo):
    lat_centroid, lng_centroid, lat_offset, lng_offset = gh.decode_exactly(geo)
    corner_1 = (lat_centroid - lat_offset, lng_centroid - lng_offset)[::-1]
    corner_2 = (lat_centroid - lat_offset, lng_centroid + lng_offset)[::-1]
    corner_3 = (lat_centroid + lat_offset, lng_centroid + lng_offset)[::-1]
    corner_4 = (lat_centroid + lat_offset, lng_centroid - lng_offset)[::-1]
    return geometry.Polygon([corner_1, corner_2, corner_3, corner_4, corner_1])
density['geometry'] = density['hash'].apply(geohash_to_polygon)

# Create a leaflet map
density['value'] = density['count'].apply(math.log) 
densityjson = gpd.GeoDataFrame(density, crs="EPSG:4326")
densityjson = densityjson.to_json()

map_center = [34.4,68.7]
map = folium.Map(location=(map_center[0], map_center[1]), 
                 tiles = "Stamen Toner", zoom_start=6)
choropleth = folium.Choropleth(geo_data = densityjson,
                               name = 'choropleth',
                               data = density,
                               columns = ['hash', 'value'],
                               key_on = 'feature.properties.hash',
                               fill_color = 'Reds',
                               fill_opacity = 0.7,
                               line_opacity = 0.2,
                               legend_name = 'Density').add_to(map)
choropleth.geojson.add_child(folium.features.GeoJsonTooltip(['hash', 'count'],labels=False))
choropleth.save("mymap.html")